sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+      // NB: Elasticsearch 5.x now provides the slice API.
+      JsonObject statsJson = getStats(false);
+      JsonObject indexStats =
+          statsJson.getAsJsonObject("indices").getAsJsonObject(
+              spec.getConnectionConfiguration().getIndex()).getAsJsonObject("primaries");
+      JsonObject store = indexStats.getAsJsonObject("store");
+      return store.getAsJsonPrimitive("size_in_bytes").getAsLong();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("shard", shardPreference));
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+      return false;
+    }
+
+    @Override
+    public BoundedReader<String> createReader(PipelineOptions options) throws IOException {
+      return new BoundedElasticsearchReader(this);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public Coder<String> getDefaultOutputCoder() {
+      return StringUtf8Coder.of();
+    }
+
+    private JsonObject getStats(boolean shardLevel) throws IOException {
+      RestClient restClient = spec.getConnectionConfiguration().createClient();
+      HashMap<String, String> params = new HashMap<>();
+      if (shardLevel) {
+        params.put("level", "shards");
+      }
+      String endpoint = String.format("/%s/_stats",
+                                      spec.getConnectionConfiguration
+                                          ().getIndex());
+      Response response =
+          restClient.performRequest("GET", endpoint, params, new BasicHeader("", ""));
+      restClient.close();
+      return parseResponse(response);
+
+    }
+
+  }
+
+  private static class BoundedElasticsearchReader extends BoundedSource.BoundedReader<String> {
+
+    private final BoundedElasticsearchSource source;
+
+    private RestClient restClient;
+    private String current;
+    private String scroll;
+
+    public BoundedElasticsearchReader(BoundedElasticsearchSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      restClient = source.spec.getConnectionConfiguration().createClient();
+
+      String query = source.spec.getQuery();
+      if (query == null) {
+        query = "{\n"
+            + "  \"query\": {\n"
+            + "    \"match_all\": {}\n"
+            + "  }\n"
+            + "}";
+      }
+
+      Response response;
+      String endPoint = String.format("/%s/%s/_search",
+                                      source.spec.getConnectionConfiguration().getIndex(),
+                                      source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive() != null ? source.spec
```

Comment:
Good point. 
When using scroll API, ES needs a statefull scroll context server side. This keepalive is the delay after which it will remove the context. This delay needs to be long enough to process the chunk of data returned by previous scroll request. In our case - for start() and advance() - each chunk contains only 1 doc. So if the processing of a single doc is longer than 1 minute before next advance() - not next start() because each start() starts a new scroll - , the read will fail indeed. That is why we allow the user to specify a keepalive in case his processing of one doc is longer than default 1 minute.